cli/azd/cmd/middleware/hooks.go

```diff
@@ -0,0 +1,159 @@
+package middleware
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/azure/azure-dev/cli/azd/cmd/actions"
+	"github.com/azure/azure-dev/cli/azd/pkg/environment"
+	"github.com/azure/azure-dev/cli/azd/pkg/exec"
+	"github.com/azure/azure-dev/cli/azd/pkg/ext"
+	"github.com/azure/azure-dev/cli/azd/pkg/input"
+	"github.com/azure/azure-dev/cli/azd/pkg/project"
+)
+
+type HooksMiddleware struct {
+	env           *environment.Environment
+	projectConfig *project.ProjectConfig
+	commandRunner exec.CommandRunner
+	console       input.Console
+	options       *Options
+}
+
+// Creates a new instance of the Hooks middleware
+func NewHooksMiddleware(
+	env *environment.Environment,
+	projectConfig *project.ProjectConfig,
+	commandRunner exec.CommandRunner,
+	console input.Console,
+	options *Options,
+) Middleware {
+	return &HooksMiddleware{
+		env:           env,
+		projectConfig: projectConfig,
+		commandRunner: commandRunner,
+		console:       console,
+		options:       options,
+	}
+}
+
+// Runs the Hooks middleware
+func (m *HooksMiddleware) Run(ctx context.Context, next NextFn) (*actions.ActionResult, error) {
+	if err := m.registerServiceHooks(ctx); err != nil {
+		return nil, fmt.Errorf("failed registering service hooks, %w", err)
+	}
+
+	return m.registerCommandHooks(ctx, next)
+}
+
+// Register command level hooks for the executing cobra command & action
+// Invokes the middleware next function
+func (m *HooksMiddleware) registerCommandHooks(ctx context.Context, next NextFn) (*actions.ActionResult, error) {
+	if m.projectConfig.Hooks == nil || len(m.projectConfig.Hooks) == 0 {
+		log.Println("project does not contain any command hooks.")
+		return next(ctx)
+	}
+
+	hooksManager := ext.NewHooksManager(m.projectConfig.Path)
+	hooksRunner := ext.NewHooksRunner(
+		hooksManager,
+		m.commandRunner,
+		m.console,
+		m.projectConfig.Path,
+		m.projectConfig.Hooks,
+		m.env.Environ(),
+	)
+
+	var actionResult *actions.ActionResult
+
+	commandNames := []string{m.options.Name}
+	commandNames = append(commandNames, m.options.Aliases...)
+
+	err := hooksRunner.Invoke(ctx, commandNames, func() error {
+		result, err := next(ctx)
+		if err != nil {
+			return err
+		}
+
+		actionResult = result
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return actionResult, nil
+}
+
+// Registers event handlers for all services within the project configuration
+// Runs hooks for each matching event handler
+func (m *HooksMiddleware) registerServiceHooks(ctx context.Context) error {
+	for serviceName, service := range m.projectConfig.Services {
+		// If the service hasn't configured any hooks we can continue on.
+		if service.Hooks == nil || len(service.Hooks) == 0 {
+			log.Printf("service '%s' does not require any command hooks.\n", serviceName)
```

Comment:
I thought it would be good data to have for debugging when someone is using command hooks.  Happy to remove if you don't think it provides much value.